File: uwp/CoppaLib/Coppa.cs
/**
 Copyright 2016 Just10 Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
 
using System;

namespace Just10
{
    /// <summary>
    /// Provides static COPPA compliancy age-check methods
    /// </summary>
    public static class COPPA
    {
        private const int MINIMUM_AGE = 14;

        /// <summary>
        /// Determines if the provided birthdate provides an age that satisfies COPPA requirements
        /// </summary>
        /// <param name="birthdate">Birthdate as DateTime instance</param>
        /// <returns>True if the calculated age meets COPPA requirements &amp; False otherwise</returns>
        public static bool AgeMeetsCOPPARequirements (DateTime birthdate)
        {
            if (birthdate == null)
                throw new ArgumentNullException ("birthdate");

            return birthdate.GetAge () >= MINIMUM_AGE;
        }

        /// <summary>
        /// Determines if the provided birthdate provides an age that satisfies COPPA requirements
        /// </summary>
        /// <param name="birthdate">Birthdate as Unix timestamp in UTC</param>
        /// <returns>True if the calculated age meets COPPA requirements &amp; False otherwise</returns>
        public static bool AgeMeetsCOPPARequirements (long birthdate)
        {
            return AgeMeetsCOPPARequirements(DateTimeExtensions.FromMillis (birthdate));
        }

        /// <summary>
        /// Determines if the provided birthdate provides an age that satisfies COPPA requirements
        /// </summary>
        /// <param name="yyyymmdd">Birthdate as a YYYYMMDD date string</param>
        /// <returns>True if the calculated age meets COPPA requirements &amp; False otherwise</returns>
        public static bool AgeMeetsCOPPARequirements (string yyyymmdd)
        {
            return AgeMeetsCOPPARequirements(DateTimeExtensions.FromString(yyyymmdd));
        }

    }
}

File: bb10/COPPA/src/coppa.cpp
/**
 Copyright 2016 Just10 Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
 
 /*
 * coppa.cpp
 *
 *  Created on: May 31, 2016
 *      Author: mahramf
 */

#include <coppa.h>
#include <time_utils.h>
#include <QtCore/QDateTime>
#include <QtCore/QString>

namespace just10
{
    namespace coppa
    {
        const int MINIMUM_AGE = 14;

        bool ageMeetsCOPPARequirements(const QString &birthdate)
        {
            return ageMeetsCOPPARequirements(util::fromYYYYMMDD(birthdate));
        }

        bool ageMeetsCOPPARequirements(const QDate &birthdate)
        {
            return util::getAge(birthdate) >= MINIMUM_AGE;
        }

        bool ageMeetsCOPPARequirements(const qint64 birthdate)
        {
            return ageMeetsCOPPARequirements(QDateTime::fromMSecsSinceEpoch(birthdate).date());
        }

        bool ageMeetsCOPPARequirements(const QDateTime &birthdate)
        {
            return ageMeetsCOPPARequirements(birthdate.date());
        }
    }
}

